### ISSUE TITLE
docker host dns (- PEERDNS=auto)  not working

### ISSUE THREAD

**Issue by hereisderek** [linuxserver/docker-wireguard#106]
[linuxserverurl]: https://linuxserver.io
[![linuxserver.io](https://raw.githubusercontent.com/linuxserver/docker-templates/master/linuxserver.io/img/linuxserver_medium.png)][linuxserverurl]

------------------------------

## Expected Behavior
DNS lookup on the client side should work when point dns server to the docker's IP (e.g. 10.13.13.1)

## Current Behavior
DNS lookup timeout
```
➜  ~ nslookup google.com 10.13.13.1
;; connection timed out; no servers could be reached
```

## Steps to Reproduce
as documented 

## Environment
my docker host machine has both working ipv4/ipv6 addresses,
the client machine i used to connect to the wireguard has just ipv4

docker-compose.yml

```yaml

version: "2.1"
services:
  wireguard:
    image: linuxserver/wireguard
    container_name: wireguard
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=Pacific/Auckland
      - SERVERURL=######optional
      - SERVERPORT=500 #optional
      - PEERS=mac
      - PEERDNS=auto 
      # however, if I use the following dns on the client side it's working
      # - PEERDNS=2606:4700:4700::1111,2606:4700:4700::1001,1.1.1.1,1.0.0.1,8.8.8.8,8.8.4.4
      - INTERNAL_SUBNET=10.13.13.0 #optional
      - ALLOWEDIPS=0.0.0.0/0 #optional
    volumes:
      - ./config:/config
      - /lib/modules:/lib/modules
    ports:
      - 500:51820/udp
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
      - "net.ipv6.conf.all.disable_ipv6=0"
      - "net.ipv6.conf.all.forwarding=1"
      - "net.ipv6.conf.default.forwarding=1"
    restart: unless-stopped

```

docker network
```
derek in /srv/docker/wireguard  > docker network ls
NETWORK ID     NAME                DRIVER    SCOPE
c5c068e450a2   bridge              bridge    local
28a302895ed2   host                host      local
e7a25adbf216   none                null      local
1cbd707b8e79   rclone_network      bridge    local
18f32be7c6ea   wireguard_default   bridge    local
derek in /srv/docker/wireguard  > docker network inspect wireguard_default
[
    {
        "Name": "wireguard_default",
        "Id": "18f32be7c6ea67b4b14b2e52ac9ee5c0fc256506f194155a9890b76e5d0b3b55",
        "Created": "2021-04-20T14:13:52.305755287+12:00",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.22.0.0/16",
                    "Gateway": "172.22.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": true,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "a5b64ab30317235854fa6144124e2a76046237ed9c44a88feb48b214fca7e01a": {
                "Name": "wireguard",
                "EndpointID": "4caa55a615aa915d02aba8fc3b59ec22e9d0c054ca325b115fa6e917fbf75088",
                "MacAddress": "02:42:ac:16:00:02",
                "IPv4Address": "172.22.0.2/16",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {
            "com.docker.compose.network": "default",
            "com.docker.compose.project": "wireguard",
            "com.docker.compose.version": "1.29.1"
        }
    }
]
derek in /srv/docker/wireguard  >
```

with generated client config:

```yaml
[Interface]
PrivateKey = ***
ListenPort = 51820
Address = 10.13.13.2/32
DNS = 10.13.13.1

[Peer]
PublicKey = ***
AllowedIPs = 0.0.0.0/0
Endpoint = ***

```
I wasn't able to resolve any dns query. running on client machine (that's connected to the wireguard server)

```
➜  ~ nslookup google.com 10.13.13.1
;; connection timed out; no servers could be reached

➜  ~ nslookup google.com 1.1.1.1
Server:		1.1.1.1
Address:	1.1.1.1#53

Non-authoritative answer:
Name:	google.com
Address: 142.250.66.174
``` 

running on the docker's host machine 

```
derek in /srv/docker/wireguard  > nslookup google.com
Server:		127.0.0.53
Address:	127.0.0.53#53

Non-authoritative answer:
Name:	google.com
Address: 216.58.199.78
Name:	google.com
Address: 2404:6800:4006:806::200e

derek in /srv/docker/wireguard  > cat /etc/resolv.conf
# This file is managed by man:systemd-resolved(8). Do not edit.
#
# This is a dynamic resolv.conf file for connecting local clients to the
# internal DNS stub resolver of systemd-resolved. This file lists all
# configured search domains.
#
# Run "resolvectl status" to see details about the uplink DNS servers
# currently in use.
#
# Third party programs should typically not access this file directly, but only
# through the symlink at /etc/resolv.conf. To manage man:resolv.conf(5) in a
# different way, replace this symlink by a static file or a different symlink.
#
# See man:systemd-resolved.service(8) for details about the supported modes of
# operation for /etc/resolv.conf.

nameserver 127.0.0.53
options edns0 trust-ad
```

dns is also working in the wireguard image
```
derek in /srv/docker/wireguard  > docker exec -it wireguard  sh
# cat /etc/resolv.conf
nameserver 127.0.0.11
options edns0 trust-ad ndots:0
# ping google.com
PING google.com (216.58.199.78) 56(84) bytes of data.
64 bytes from syd15s01-in-f78.1e100.net (216.58.199.78): icmp_seq=1 ttl=118 time=25.1 ms
64 bytes from syd15s01-in-f78.1e100.net (216.58.199.78): icmp_seq=2 ttl=118 time=24.7 ms
64 bytes from syd15s01-in-f78.1e100.net (216.58.199.78): icmp_seq=3 ttl=118 time=25.1 ms
^C
--- google.com ping statistics ---
3 packets transmitted, 3 received, 0% packet loss, time 2003ms
rtt min/avg/max/mdev = 24.766/25.026/25.172/0.184 ms

root@ef161b66f634:/# cat /etc/resolv.conf
nameserver 127.0.0.11
nameserver 2606:4700:4700::1111
nameserver 2606:4700:4700::1001
options edns0 trust-ad ndots:0

# ip a
1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
    link/loopback 00:00:00:00:00:00 brd 00:00:00:00:00:00
    inet 127.0.0.1/8 scope host lo
       valid_lft forever preferred_lft forever
    inet6 ::1/128 scope host
       valid_lft forever preferred_lft forever
3: wg0: <POINTOPOINT,NOARP,UP,LOWER_UP> mtu 1420 qdisc noqueue state UNKNOWN group default qlen 1000
    link/none
    inet 10.13.13.1/32 scope global wg0
       valid_lft forever preferred_lft forever
29: eth0@if30: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
    link/ether 02:42:ac:16:00:02 brd ff:ff:ff:ff:ff:ff link-netnsid 0
    inet 172.22.0.2/16 brd 172.22.255.255 scope global eth0
       valid_lft forever preferred_lft forever
    inet6 fe80::42:acff:fe16:2/64 scope link
       valid_lft forever preferred_lft forever

root@0abcb9f41e19:/# iptables -S
-P INPUT ACCEPT
-P FORWARD ACCEPT
-P OUTPUT ACCEPT
-A FORWARD -i wg0 -j ACCEPT
-A FORWARD -o wg0 -j ACCEPT
root@0abcb9f41e19:/# iptables -L -v -n
Chain INPUT (policy ACCEPT 17300 packets, 51M bytes)
 pkts bytes target     prot opt in     out     source               destination

Chain FORWARD (policy ACCEPT 0 packets, 0 bytes)
 pkts bytes target     prot opt in     out     source               destination
    0     0 ACCEPT     all  --  wg0    *       0.0.0.0/0            0.0.0.0/0
    0     0 ACCEPT     all  --  *      wg0     0.0.0.0/0            0.0.0.0/0

Chain OUTPUT (policy ACCEPT 8273 packets, 470K bytes)
 pkts bytes target     prot opt in     out     source               destination

root@ef161b66f634:/#  nslookup google.com 10.13.13.1 # after installing dnsutils
Server:		10.13.13.1
Address:	10.13.13.1#53

Non-authoritative answer:
Name:	google.com
Address: 216.58.199.78
Name:	google.com
Address: 2404:6800:4006:80a::200e

```

## Docker logs

initial wireguard logs

```
derek in /srv/docker/wireguard  > docker logs -f wireguard
[s6-init] making user provided files available at /var/run/s6/etc...exited 0.
[s6-init] ensuring user provided files have correct perms...exited 0.
[fix-attrs.d] applying ownership & permissions fixes...
[fix-attrs.d] done.
[cont-init.d] executing container initialization scripts...
[cont-init.d] 01-envfile: executing...
[cont-init.d] 01-envfile: exited 0.
[cont-init.d] 10-adduser: executing...

-------------------------------------
          _         ()
         | |  ___   _    __
         | | / __| | |  /  \
         | | \__ \ | | | () |
         |_| |___/ |_|  \__/


Brought to you by linuxserver.io
-------------------------------------

To support the app dev(s) visit:
WireGuard: https://www.wireguard.com/donations/

To support LSIO projects visit:
https://www.linuxserver.io/donate/
-------------------------------------
GID/UID
-------------------------------------

User uid:    1000
User gid:    1000
-------------------------------------

[cont-init.d] 10-adduser: exited 0.
[cont-init.d] 30-config: executing...
Uname info: Linux a5b64ab30317 5.11.13-051113-generic #202104101231 SMP Sat Apr 10 12:34:46 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux
**** It seems the wireguard module is already active. Skipping kernel header install and module compilation. ****
**** Server mode is selected ****
**** External server address is set to ##### ****
**** External server port is set to 500. Make sure that port is properly forwarded to port 51820 inside this container ****
**** Internal subnet is set to 10.13.13.0 ****
**** AllowedIPs for peers 0.0.0.0/0 ****
**** PEERDNS var is either not set or is set to "auto", setting peer DNS to 10.13.13.1 to use wireguard docker host's DNS. ****
**** Server mode is selected ****
**** Server related environment variables changed, regenerating 1 server and mac

## QR images omitted ##

[cont-init.d] 30-config: exited 0.
[cont-init.d] 99-custom-scripts: executing...
[custom-init] no custom files found exiting...
[cont-init.d] 99-custom-scripts: exited 0.
[cont-init.d] done.
[services.d] starting services
[services.d] done.
[#] ip link add wg0 type wireguard
[#] wg setconf wg0 /dev/fd/63
[#] ip -4 address add 10.13.13.1 dev wg0
[#] ip link set mtu 1420 up dev wg0
[#] ip -4 route add 10.13.13.9/32 dev wg0
[#] ip -4 route add 10.13.13.8/32 dev wg0
[#] ip -4 route add 10.13.13.7/32 dev wg0
[#] ip -4 route add 10.13.13.6/32 dev wg0
[#] ip -4 route add 10.13.13.5/32 dev wg0
.:53
CoreDNS-1.8.3
linux/amd64, go1.16, 4293992
[#] ip -4 route add 10.13.13.4/32 dev wg0
[#] ip -4 route add 10.13.13.3/32 dev wg0
[#] ip -4 route add 10.13.13.2/32 dev wg0
[#] ip -4 route add 10.13.13.11/32 dev wg0
[#] ip -4 route add 10.13.13.10/32 dev wg0
[#] iptables -A FORWARD -i wg0 -j ACCEPT; iptables -A FORWARD -o wg0 -j ACCEPT; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE

```

## UPDATE

I am able to get dns through wiregard working by editing `Corefile` content to
```
. {
    forward . 1.1.1.1 8.8.8.8 9.9.9.9 
    log
    errors
}
```

however, still not sure why forwarding to the host didn't work

## UPDATE #2 
while running inside of wireguard image

```
derek in /srv/docker/wireguard  > docker exec -it wireguard  bash
root@ef161b66f634:/# cat /etc/resolv.conf
nameserver 127.0.0.11
nameserver 2606:4700:4700::1111
nameserver 2606:4700:4700::1001
options edns0 trust-ad ndots:0
```

**Comment by bananflugan** [comment]
I have the same issue, using my docker host´s DNS will not work.

**Comment by hereisderek** [comment]
bump

**Comment by LockedThread** [comment]
same issue over here too

**Comment by bananflugan** [comment]
I got my DNS working a while back which was logged in a different issue. Here is my config.

```
wireguard:
    image: ghcr.io/linuxserver/wireguard
    container_name: wireguard
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      - PUID=1003
      - PGID=1003
      - TZ=Europe/Stockholm
      - SERVERURL=wireguard.DOMAIN.COM #optional
      - SERVERPORT=51820 #optional
      - PEERS=my_iphone, wife_iphone #optional
      - PEERDNS=auto #optional
      - INTERNAL_SUBNET=10.13.13.0 #optional
      - ALLOWEDIPS=192.168.10.0/24, 10.13.13.1 #only using the VPN for traffic to my LAN
    volumes:
      - /data/docker/wireguard/config:/config
      - /lib/modules:/lib/modules
    ports:
      - 51820:51820/udp
    #sysctls:
      #- net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped
    dns:
      - 192.168.100.14 #ip of my pi-hole container (this setting is important)
    networks:
      docker_net:
        ipv4_address: 192.168.100.16
```

**Comment by LockedThread** [comment]
> I got my DNS working a while back which was logged in a different issue. Here is my config.
> 
> ```
> wireguard:
>     image: ghcr.io/linuxserver/wireguard
>     container_name: wireguard
>     cap_add:
>       - NET_ADMIN
>       - SYS_MODULE
>     environment:
>       - PUID=1003
>       - PGID=1003
>       - TZ=Europe/Stockholm
>       - SERVERURL=wireguard.DOMAIN.COM #optional
>       - SERVERPORT=51820 #optional
>       - PEERS=my_iphone, wife_iphone #optional
>       - PEERDNS=auto #optional
>       - INTERNAL_SUBNET=10.13.13.0 #optional
>       - ALLOWEDIPS=192.168.10.0/24, 10.13.13.1 #only using the VPN for traffic to my LAN
>     volumes:
>       - /data/docker/wireguard/config:/config
>       - /lib/modules:/lib/modules
>     ports:
>       - 51820:51820/udp
>     #sysctls:
>       #- net.ipv4.conf.all.src_valid_mark=1
>     restart: unless-stopped
>     dns:
>       - 192.168.100.14 #ip of my pi-hole container (this setting is important)
>     networks:
>       docker_net:
>         ipv4_address: 192.168.100.16
> ```

What I ended up doing is putting wireguard and pihole on the same network and assigned them static ip address to reference each other. I was just being stupid and this was a me problem.

**Comment by hereisderek** [comment]
bump

**Comment by LockedThread** [comment]
bump

**Comment by taiyodayo** [comment]
>     dns:
>       - 192.168.100.14 #ip of my pi-hole container (this setting is important)

@bananflugan 
thank you for the pointer!
I've long had this issue of my wg clients not being able to use wg server host's dns.

In may case I simply added docker host's address as command line argument:
```
  --dns=172.17.0.1 \
```

and all is working great now!
( I have dnsmasq setup on the host.)

**Comment by reinisb** [comment]
same issue

**Comment by Zauberio** [comment]
same issue...

**Comment by caheredia** [comment]
I've tried many permutations of `PEERDNS` and `dns` in docker compose with no success. Unfortunately, same issue

**Comment by instantdreams** [comment]
Having the same issue - tunnel connected, no dns resolution.

## Environment
Raspberry Pi 4 2GB running nginx proxy manager, pihole, duckdns, and wireguard
Duplicate Pi running the same containers
Router points to both Pis as local DNS
Router port forwards 51820, 80, 443 to primary Pi (WireGuard, NPM)
General configuration is solid


### docker-compose.yml
```
version: "3"
services:
  wireguard:
    image: lscr.io/linuxserver/wireguard
    container_name: wireguard
    hostname: ${HOSTNAME}
    networks:
      - wireguard_net
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    ports:
      - 51820:51820/udp
    dns:
      - 192.168.1.1
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
      - TZ=${TZ}
      - SERVERURL=${SERVERURL}
      - SERVERPORT=${SERVERPORT}
      - PEERS=${PEERS}
      - PEERDNS=${PEERDNS}
      - INTERNAL_SUBNET=${SUBNET}
      - ALLOWEDIPS=${ALLOWEDIPS}
    volumes:
      - ${APPDATA}:/config
      - /lib/modules:/lib/modules
    restart: unless-stopped
networks:
  wireguard_net:
    driver: bridge
    name: wireguard_net
```
### .env
```
# User ID and Group ID
PUID=1000
PGID=1000
# Timezone from https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
TZ=[timezone]
# Directory locations
APPDATA=/srv/wireguard/config/
# Container Specifics
HOSTNAME=pihole1
SERVERURL=[subdomain]
SERVERPORT=51820
PEERS=1
PEERDNS=192.168.1.1 # router ip
SUBNET=10.13.13.0
ALLOWEDIPS=0.0.0.0/0
```

### docker network
```
 $ docker network ls
NETWORK ID     NAME                          DRIVER    SCOPE
c9bb91b727df   bridge                        bridge    local
43a67469ffce   duckdns_default               bridge    local
08df0f4d3af4   host                          host      local
0b5fa9e3a8c8   nginx-proxy-manager_default   bridge    local
72b53c9b6bb6   none                          null      local
a6cb02fc92bd   pihole_default                bridge    local
8a2fcc0cfae9   wireguard_net                 bridge    local
```

### wireguard_net details
```
$ docker network inspect wireguard_net
[
    {
        "Name": "wireguard_net",
        "Id": "8a2fcc0cfae98c3f3ed2d281f21b521b6e0e0df095c8af31c3a912c2fb392c89",
        "Created": "2022-01-13T08:03:51.452454546-07:00",
        "Scope": "local",
        "Driver": "bridge",
        "EnableIPv6": false,
        "IPAM": {
            "Driver": "default",
            "Options": null,
            "Config": [
                {
                    "Subnet": "172.30.0.0/16",
                    "Gateway": "172.30.0.1"
                }
            ]
        },
        "Internal": false,
        "Attachable": true,
        "Ingress": false,
        "ConfigFrom": {
            "Network": ""
        },
        "ConfigOnly": false,
        "Containers": {
            "9c88bf347f3b917a86ba1150c73e835fe166fa5e14b4d14066d006da8ecf8038": {
                "Name": "wireguard",
                "EndpointID": "b06d32196e76d527940d451e847fc0388b29dbf5595f7e389b8bc34ba110c841",
                "MacAddress": "02:42:ac:1e:00:02",
                "IPv4Address": "172.30.0.2/16",
                "IPv6Address": ""
            }
        },
        "Options": {},
        "Labels": {
            "com.docker.compose.network": "wireguard_net",
            "com.docker.compose.project": "wireguard",
            "com.docker.compose.version": "1.29.2"
        }
    }
]
```

### wg0.conf
```
[Interface]
Address = 10.13.13.1
ListenPort = 51820
PrivateKey = *
PostUp = iptables -A FORWARD -i %i -j ACCEPT; iptables -A FORWARD -o %i -j ACCEPT; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
PostDown = iptables -D FORWARD -i %i -j ACCEPT; iptables -D FORWARD -o %i -j ACCEPT; iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE

[Peer]
# peer1
PublicKey = *
AllowedIPs = 10.13.13.2/32
```

### peer1.conf
```
[Interface]
Address = 10.13.13.2
PrivateKey = *
ListenPort = 51820
DNS = 192.168.1.1

[Peer]
PublicKey = *
Endpoint = [subdomain]:51820
AllowedIPs = 0.0.0.0/0
```

Any suggestions?

**Comment by aptalca** [comment]
Do you see the handshake on the server? `docker exec -it wireguard wg show`

**Comment by instantdreams** [comment]
I do:
```
$ docker exec -it wireguard wg show
interface: wg0
  public key: [key]
  private key: (hidden)
  listening port: 51820

peer: [key]
  endpoint: [ip]:51820
  allowed ips: 10.13.13.2/32
  latest handshake: 59 seconds ago
  transfer: 213.38 KiB received, 573.84 KiB sent
```

But also... It seems to be working now. Will test further!

**Comment by instantdreams** [comment]
> Will test further!

Just tested the following two clients:
1. Microsoft Surface Go 2 running Windows 11 Pro v22000.434 and WireGuard Windows app v0.5.3
2. Google Pixel 6 running  Android 12 v5.10.43 and WireGuard 1.0.202111029

They are both connecting and resolving DNS fine but I'm currently on my home network. Will do an extended test away from home later to check if they can resolve local addresses and internet sites.

**Comment by instantdreams** [comment]
Just tested the following two clients:
1. Microsoft Surface Go 2 running Windows 11 Pro v22000.434 and WireGuard Windows app v0.5.3
2. Google Pixel 6 running  Android 12 v5.10.43 and WireGuard 1.0.202111029

They are both connecting but **not resolving** DNS at all.

**Comment by instantdreams** [comment]
Could it be an issue with resolv.conf within the wireguard container?
```
$ docker exec -it wireguard cat /etc/resolv.conf
nameserver 127.0.0.11
options ndots:0
```

I would expect that to be either at least 127.0.0.1 or by preference the 192.168.1.1 I pass in to PEERDNS.

**Comment by aptalca** [comment]
PEERDNS has no effect on the server. As the name suggests, it only sets the DNS value in the peer conf, which you can override in the peer.
resolv.conf content is determined by docker, `127.0.0.11` is docker's internal dns

**Comment by instantdreams** [comment]
Then I'm stumped. My set up appears to be connecting without issue:
```
WireGuard Client on Android Phone -> My IP Address:My Port -> Router
Router -> Port Forward to Server
Server -> Docker Container with WireGuard
```

WireGuard Server reports the handshake is good. WireGuard Client reports a solid connection in the logs. The issue is the DNS just doesn't resolve.

Using Docker Host networking won't fix this. Futzing with the docker container settings is my only hope.

Here are the relevant portions of my WAN configuration:
```
Router:           192.168.1.1
Subnet:           255.255.255.0 
Router DNS 1:     192.168.1.2
Router DNS 2:     192.168.1.3
Pihole 1:         192.168.1.2
Pihole 2:         192.168.1.3
```

And the WireGuard Server container:
```
Internal IP:      10.13.13.1
Internal Subnet:  10.13.13.0
Allowed IPs:      0.0.0.0/0
Peer DNS:         192.168.1.1
Peer Allowed IPs: 10.13.13.2/32
```

And the WireGuard Server Peer:
```
Peer1 Address:     10.13.13.2
Peer 1 DNS:        192.168.1.1
Peer 1 Allowed IPs: 0.0.0.0/0
```

I'm so deep into this now that I probably can't see the forest for the trees. Anyone see something obvious I've missed?

**Comment by aptalca** [comment]
You pointed your router's dns to pihole. Is your router running a dns forwarder? If not, `192.168.1.1` is not a valid entry for peerdns.

If your router is using pihole and is set up with a forwarder, then you don't need to do any extra steps like setting the dns in compose, or manually setting a PEERDNS value.

Docker should automatically send the dns requests to your router, which should forward to pihole. Wireguard container will use the docker dns settings in `auto`.

**Comment by caheredia** [comment]
| **Ubuntu VM** |
|---------------|
| Docker Compose       |
| - pi hole     |
| - wireguard   |

This is my stack. Through wireguard I can connect to my internal network just fine, but only with internal ip addresses not lan names—for example 192.1.68.1.1 works. My router is set to use pi hole for DNS. However, wireguard connections are not using pi-hole. I suspect this is probably due to the VM/host using it's own DNS. The VM not using the pi-hole for DNS makes sense; otherwise when pi-hole is down I wouldn't be able to update the VM. 

My question/issue (related to others above): How do I get wireguard to use pi-hole for DNS when it's running in the same stack/VM as pi-hole?

**Comment by aptalca** [comment]
Set the dns in the client conf to the pihole ip

**Comment by caheredia** [comment]
> Set the dns in the client conf to the pihole ip

If I do that, then wireguard doesn't connect to anything other than the wireguard server.

**Comment by aptalca** [comment]
I don't follow. Point the client dns to whatever dns server you want to use, whether it's 8.8.8.8, 1.1.1.1 or the pihole that's local to the wireguard server.

You provided very little info on your infra, it's impossible for us to troubleshoot any further.

**Comment by caheredia** [comment]
> I don't follow. Point the client dns to whatever dns server you want to use, whether it's 8.8.8.8, 1.1.1.1 or the pihole that's local to the wireguard server.

Thanks for the quick reply. As a fellow software developer, I very much appreciate the effort you all put into this project. 

As I stated above, when I use my pi-hole as DNS in client configs I can't connect to the internet anymore. 

My configs: 
```docker
version: '3'

services:

# More info at https://github.com/pi-hole/docker-pi-hole/ and https://docs.pi-hole.net/
  pihole:
    container_name: pihole
    hostname: pi-hole
    image: pihole/pihole:latest
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "80:80/tcp"
    environment:
      TZ: 'America/Tijuana'
      WEBPASSWORD: ${PIHOLE_PASSWORD} 
      ServerIP: 192.168.1.214
    # Volumes store your data between container upgrades
    volumes:
      - ./pihole/etc-pihole/:/etc/pihole/
      - ./pihole/etc-dnsmasq.d/:/etc/dnsmasq.d/
    # Recommended but not required (DHCP needs NET_ADMIN)
    #   https://github.com/pi-hole/docker-pi-hole#note-on-capabilities
    #    cap_add:
    #  - NET_ADMIN
    restart: unless-stopped

  wireguard:
    image: lscr.io/linuxserver/wireguard
    container_name: wireguard
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      TZ: 'America/Tijuana'
      PUID: 1000
      PGID: 1000
      SERVERURL: ${SERVERURL} #optional
      SERVERPORT: 51820 #optional
      PEERS: "iphone, macbook"
      ALLOWEDIPS: 0.0.0.0/0 #optional
    volumes:
      - /home/ubuntu/wireguard/config:/config
      - /lib/modules:/lib/modules
    ports:
      - 51820:51820/udp
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped
```
Example Client config, this works but doesn't use pi-hole even though router is using pi-hole for DNS
```
[Interface]
PrivateKey = some_key
ListenPort = 51820
Address = 10.13.13.3/32
DNS = 10.13.13.1

[Peer]
PublicKey = some_key
AllowedIPs = 0.0.0.0/0
Endpoint = endpoint:51820
```
Pointing client configs to the pi-hole results in a loss of internet connection 
```
[Interface]
PrivateKey = some_key
ListenPort = 51820
Address = 10.13.13.3/32
DNS = 192.168.1.214

[Peer]
PublicKey = some_key
AllowedIPs = 0.0.0.0/0
Endpoint = endpoint:51820
```

**Comment by aptalca** [comment]
It's not that you can't connect to anything else. The issue is, you can't connect to pihole, because pihole is running in macvlan whereas wireguard server is running in user defined bridge. Macvlan is not accessible from the host or bridge networks on the host, which is a docker security feature.

**Comment by aptalca** [comment]
Closing this thread as it veered way off of the original topic. These issues are not related to potential bugs with the `PEERDNS=auto` feature, which is working as intended.

**Comment by caheredia** [comment]
> ~It's not that you can't connect to anything else. The issue is, you can't connect to pihole, because pihole is running in macvlan whereas wireguard server is running in user defined bridge. Macvlan is not accessible from the host or bridge networks on the host, which is a docker security feature.~ I misread, pihole seems to be in user defined bridge as well I'm assuming `192.168.1.124` is the lan ip of the server hosting the docker containers? Is there ufw or another firewall?

That is correct, `192.168.1.124` is the host address. No firewall is present. 
```shell
ubuntu@homelab:~$ sudo ufw status
Status: inactive
```

**Comment by aptalca** [comment]
@caheredia just did some tests
I set up pihole on my vps the same way you did, in the same compose yaml as wireguard and indeed the dns queries don't work when the vps public ip is defined as the dns address.

However, the issue is not specific to wireguard or docker.

On the same vps, if I do `dig @127.0.0.1 google.com`, pihole responds as expected.
But if I do `dig @PUBLICIP google.com` it times out.
If on another machine (over the wan) I do the same thing with the public ip, it works.

So the issue is that dns queries via external/public ip on the same machine don't work. Since you're using the external IP, that's why it's not working for you. If you try to set dns to pihole's docker network ip, it should work. But the downside is, that ip by default is not static and may change when the containers are recreated. You can try assigning a static ip for pihole (make sure you don't do macvlan by mistake, which won't work, but instead do a static ip in the user defined docker bridge network).

I have no idea why the dns queries by public ip fail on the same machine, probably some loopback failure.

**Comment by instantdreams** [comment]
> Closing this thread as it veered way off of the original topic. These issues are not related to potential bugs with the `PEERDNS=auto` feature, which is working as intended.

Indeed. I reset my container to use `PEERDNS=auto` and it is resolving external DNS correctly now. What it's not doing is resolving internal DNS, but it is attempting to do so.

Here's my configuration to spin up wireguard:
### .env
```
# User ID and Group ID
PUID=1000
PGID=1000
# Timezone from https://en.wikipedia.org/wiki/List_of_tz_database_time_zones
TZ=America/Edmonton
# Directory locations
APPDATA=/srv/wireguard/config/
# Container Specifics
HOSTNAME=[hostname]
SERVERURL=[server domain]
SERVERPORT=51820
PEERS=1
PEERDNS=auto
SUBNET=10.13.13.0
ALLOWEDIPS=0.0.0.0/0
```

### docker-compose.yml
```
version: "3"
services:
  wireguard:
    image: lscr.io/linuxserver/wireguard
    container_name: wireguard
    hostname: ${HOSTNAME}
    networks:
      - wireguard_net
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    ports:
      - 51820:51820/udp
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
      - TZ=${TZ}
      - SERVERURL=${SERVERURL}
      - SERVERPORT=${SERVERPORT}
      - PEERS=${PEERS}
      - PEERDNS=${PEERDNS}
      - INTERNAL_SUBNET=${SUBNET}
      - ALLOWEDIPS=${ALLOWEDIPS}
    volumes:
      - ${APPDATA}:/config
      - /lib/modules:/lib/modules
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped
networks:
  wireguard_net:
    driver: bridge
    name: wireguard_net
```

Before connecting to the Wireguard tunnel:
```
PS C:\> ping www.google.com

Pinging www.google.com [2607:f8b0:400a:806::2004] with 32 bytes of data:
Reply from 2607:f8b0:400a:806::2004: time=23ms
Reply from 2607:f8b0:400a:806::2004: time=25ms
Reply from 2607:f8b0:400a:806::2004: time=25ms
Reply from 2607:f8b0:400a:806::2004: time=25ms

Ping statistics for 2607:f8b0:400a:806::2004:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 23ms, Maximum = 25ms, Average = 24ms
```

After connecting to wireguard:
```
PS C:\> ping www.google.com

Pinging www.google.com [142.251.33.68] with 32 bytes of data:
Reply from 142.251.33.68: bytes=32 time=33ms TTL=117
Reply from 142.251.33.68: bytes=32 time=35ms TTL=117
Reply from 142.251.33.68: bytes=32 time=29ms TTL=117
Reply from 142.251.33.68: bytes=32 time=29ms TTL=117

Ping statistics for 142.251.33.68:
    Packets: Sent = 4, Received = 4, Lost = 0 (0% loss),
Approximate round trip times in milli-seconds:
    Minimum = 29ms, Maximum = 35ms, Average = 31ms
```

So, the external DNS is working a treat! The tunnel uses ipv4 rather than ipv6, which is interesting to note.

However, internal DNS is not resolving correctly:
```
PS C:\> ping 192.168.1.100

Pinging 192.168.1.100 with 32 bytes of data:
General failure.
General failure.
General failure.
General failure.

Ping statistics for 192.168.1.100:
    Packets: Sent = 4, Received = 0, Lost = 4 (100% loss),
PS C:\> ping [server name]

Pinging [server name] [192.168.1.100] with 32 bytes of data:
General failure.
General failure.
General failure.
General failure.

Ping statistics for 192.168.1.100:
    Packets: Sent = 4, Received = 0, Lost = 4 (100% loss),
```

The tunnel was confirmed to be active:
```
$ docker exec -it wireguard wg show
interface: wg0
  public key: (hidden)
  private key: (hidden)
  listening port: 51820

peer: (hidden)
  endpoint: 104.157.58.164:51820
  allowed ips: 10.13.13.2/32
  latest handshake: 23 seconds ago
  transfer: 833.71 KiB received, 2.22 MiB sent
```

I'm so close now! The `PEERDNS=auto` fixed the external DNS, now I just need to fix the internal DNS. :)

**Comment by caheredia** [comment]
@aptalca Thank you for digging further. I can confirm, in order to use pihole for DNS wireguard clients must use pihole's Docker static IP for DNS: I setup a custom network, assigned pihole a static ip, and pointed the clients to the pihole's static IP, and it is working as expected. Many thanks! I'm officially a sponsor of the project.

```yaml
version: '3'

services:

# More info at https://github.com/pi-hole/docker-pi-hole/ and https://docs.pi-hole.net/
  pihole:
    container_name: pihole
    hostname: pi-hole
    image: pihole/pihole:latest
    ports:
      - "53:53/tcp"
      - "53:53/udp"
      - "67:67/udp"
      - "80:80/tcp"
    environment:
      TZ: 'America/Tijuana'
      WEBPASSWORD: ${PIHOLE_PASSWORD} 
      ServerIP: 192.168.1.214
    # Volumes store your data between container upgrades
    volumes:
      - ./pihole/etc-pihole/:/etc/pihole/
      - ./pihole/etc-dnsmasq.d/:/etc/dnsmasq.d/
    # Recommended but not required (DHCP needs NET_ADMIN)
    #   https://github.com/pi-hole/docker-pi-hole#note-on-capabilities
    #    cap_add:
    #  - NET_ADMIN
    restart: unless-stopped
    networks:
      app_net:
        ipv4_address: 172.19.0.10

  wireguard:
    image: lscr.io/linuxserver/wireguard
    container_name: wireguard
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      TZ: 'America/Tijuana'
      PUID: 1000
      PGID: 1000
      SERVERURL: ${SERVERURL} #optional
      SERVERPORT: 51820 #optional
      PEERS: "iphone, macbook"
      ALLOWEDIPS: 0.0.0.0/0 #optional
    volumes:
      - /home/ubuntu/homelab/wireguard/config:/config
      - /lib/modules:/lib/modules
    ports:
      - 51820:51820/udp
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped
    networks:
      - app_net
   
networks:
  app_net:
    ipam:
      driver: default
      config:
        - subnet: "172.19.0.0/24"
```
```
[Interface]
PrivateKey = some_key
ListenPort = 51820
Address = 10.13.13.3/32
DNS = 172.19.0.10

[Peer]
PublicKey = some_key
AllowedIPs = 0.0.0.0/0
Endpoint = endpoint:51820
```

**Comment by jackmusick** [comment]
Just in-case anyone else has this issue still after all of the above suggestions, the following config worked for me on the last two installs I did:

```
version: "2.1"
services:
  wireguard:
    image: lscr.io/linuxserver/wireguard
    container_name: wireguard
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/Indiana/Indianapolis
      - SERVERURL=my.domain.org
      - SERVERPORT=51820
      - PEERS=Macbook,Lenovo,iPhone
      - PEERDNS=auto
      - INTERNAL_SUBNET=10.13.13.0
      - ALLOWEDIPS=172.16.0.0/16,10.13.13.1/32
    volumes:
      - ./config:/config
      - ./libmodules:/lib/modules
    ports:
      - 51820:51820/udp
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped
```
The key was adding the Wireguard's interface IP to the ALLOWEDIPS list. After doing this, clients were able to use the Wireguard "server" IP as a forwarder to the docker host's DNS. In my home environment, this means that DNS requests go through my pi hole container with no extra work (since the host has a DHCP reservation with the pi hole's DNS). Similarly, my work environment has a DHCP reservation with the domain controller's IP for DNS. In both cases, they resolve as expected both externally and internally.

Another neat thing: If you're using Windows Server and want to resolve domains with your DNS suffix, DNS would be something like this:

`PEERDNS=10.13.13.1/32,ad.yourdomain.com`

This effectively is like auto, but adds your DNS suffix. It's worth noting that it appears at this time, the DNS suffix isn't respected on macOS, but works fine on Windows.

I'd love if someone more qualified could correct anything I've said. My guess is that it's assumed you'd need to add the Wireguard server or subnet to your allowed IP list if you're not tunneling everything with 0.0.0.0/0, but that wasn't very clear to me as someone very new to Wireguard.

As always, I appreciate the work that goes into these projects. It's really amazing what people have contributed and this is a far cry easier to setup than any other VPN in my experience.

**Comment by jamolnng** [comment]
I actually just got this working with AdGuard Home and Wireguard. here are my compose configs:

adguard-home
```yaml
version: '3'
services:
  app:
    image: 'adguard/adguardhome'
    restart: unless-stopped
    ports:
      - '53:53/tcp'
      - '53:53/udp'
      - '3000:3000/tcp'
    volumes:
      - /home/username/docker/adguard/work:/opt/adguardhome/work
      - /home/username/docker/adguard/conf:/opt/adguardhome/conf
    networks:
      adnet:
        ipv4_address: 172.19.0.10
networks:
  adnet:
    driver: bridge
    ipam:
      config:
        - subnet: "172.19.0.0/24"
```

```yaml
version: "2.1"
services:
  wireguard:
    image: lscr.io/linuxserver/wireguard
    container_name: wireguard
    cap_add:
      - NET_ADMIN
      - SYS_MODULE
    environment:
      - PUID=1000
      - PGID=1000
      - TZ=America/New_York
      - SERVERURL=wireguard.example.com #optional
      - SERVERPORT=51820 #optional
      - PEERS=Laptop,Phone #optional
      - PEERDNS=172.19.0.10 #optional
      - INTERNAL_SUBNET=10.13.13.0 #optional
      - ALLOWEDIPS=0.0.0.0/0 #optional
    volumes:
      - /home/username/docker/wireguard/config:/config
      - /lib/modules:/lib/modules
    ports:
      - 51820:51820/udp
    sysctls:
      - net.ipv4.conf.all.src_valid_mark=1
    restart: unless-stopped
    networks:
      - adguard-home_adnet
networks:
  adguard-home_adnet:
    external: true
```

**Comment by caheredia** [comment]
After some digging, this is what I've learned about Docker containers and DNS. By design, containers can not introspect their hosts, making it challenging to create a DNS service (PiHole/AdGuard) and have other containers on the same host reference that DNS service.  

For example, A Host(192.168.1.100) with DNS service (192.168.1.100:80) and Wireguard (192.168.1.100:518210) is a configuration designed for isolation, not communication. 

An alternative is to make each container/service a first-class citizen on the network without sharing any ports. Docker macvlans networks provide precisely that functionality. For containers that need to expose standard ports (80, 443, etc.) or containers that need to communicate with the Host machine, macvlans provide a viable solution. 

This is what my configuration looks like now: 
One macvlan with the same subnet as my router: 
- Host (192.168.1.100)
- DNS Service (192.168.1.101, ports :53, 80, 443)
- Wireguard (192.168.1.102, port: 518210 with DNS pointing to 192.168.1.101)
- Portainer (192.168.1.100:9000 attached to host IP and on `app_net` network)
- Nginx proxy manager (192.168.1.103, ports 80,81,443. Also using pointing to 192.168.1.101 for DNS)
     - Attaching NPM to both networks  `vlan_net` and `app_net` allows this container to reference other VMs on my network by IP and on the host machine by service name, e.g., `portainer`
```yaml
networks:
  app_net:
    ipam:
      driver: default
      config:
        - subnet: "172.19.0.0/24"

  vlan_net:
    driver: macvlan
    driver_opts:
      parent: ens18
    ipam:
      config:
        - subnet: 192.168.1.0/24
          gateway: 192.168.1.1
          ip_range: 192.168.1.101/32 # only assigning one IP here, and statically assigning `ipv4_address` in the service manifest. 
```

**Comment by makcuber** [comment]
So I discovered I had this same issue after initially thinking it was an issue with my pfSense NAT rules. Changed the Primary DNS of the Wireguard container to the docker bridge network IP of my DNS container and it worked fine, where as before wireguard kept throwing timeout errors for DNS queries from the wireguard container IP to the DNS server on the docker host IP. 

I did interestingly find however if I set the wireguard container's Primary DNS to be the **Hostname** of the DNS server, which still resolves to the docker host IP, then is works fine. So for some reason:

This DOES NOT work :: wireguard container IP >> docker host IP
But this DOES work :: wireguard container IP >> DNS server Hostname >> DNS lookup resolves to docker host IP >> docker host IP

Im no expert on docker networking, but based on my limited understanding my guess is that the intentional host isolation docker networks have prevents the port 53 DNS route that never leaves the docker host, but because the DNS lookup of the DNS hostname goes to the LAN gateway then back to the Docker host and on to the DNS container its not blocked by the isolation routing rules in docker. Using the DNS hostname works a lot better imo because if the DNS container IP changes for some reason it shouldn't affect your wireguard configuration at all, since the DNS hostname will always resolve to the Docker host IP and the lookup will just be routed to the new DNS container IP.

**Comment by ezesil** [comment]
For anyone still looking for this, i removed 0.0.0.0/0 from allowedips and it started working perfectly with just the needed networks: 

192.168.0.0/24 for LAN access and 10.13.13.0/32 for internet access through the vpn server.

**Comment by jaredtconnor** [comment]
> After some digging, this is what I've learned about Docker containers and DNS. By design, containers can not introspect their hosts, making it challenging to create a DNS service (PiHole/AdGuard) and have other containers on the same host reference that DNS service.
> 
> For example, A Host(192.168.1.100) with DNS service (192.168.1.100:80) and Wireguard (192.168.1.100:518210) is a configuration designed for isolation, not communication.
> 
> An alternative is to make each container/service a first-class citizen on the network without sharing any ports. Docker macvlans networks provide precisely that functionality. For containers that need to expose standard ports (80, 443, etc.) or containers that need to communicate with the Host machine, macvlans provide a viable solution.
> 
> This is what my configuration looks like now: One macvlan with the same subnet as my router:
> 
> * Host (192.168.1.100)
> * DNS Service (192.168.1.101, ports :53, 80, 443)
> * Wireguard (192.168.1.102, port: 518210 with DNS pointing to 192.168.1.101)
> * Portainer (192.168.1.100:9000 exposed on Host and on `app_net` network)
> * Nginx proxy manager (192.168.1.103, ports 80,81,443. Also pointing to 192.168.1.101 for DNS)
>   
>   * Attaching NPM to both networks  `vlan_net` and `app_net` allows this container to reference other VMs on my network by IP and on the host machine by service name, e.g., `portainer`
> 
> ```yaml
> networks:
>   app_net:
>     ipam:
>       driver: default
>       config:
>         - subnet: "172.19.0.0/24"
> 
>   vlan_net:
>     driver: macvlan
>     driver_opts:
>       parent: ens18
>     ipam:
>       config:
>         - subnet: 192.168.1.0/24
>           gateway: 192.168.1.1
>           ip_range: 192.168.1.101/32 # only assigning one IP here, and statically assigning `ipv4_address` in the service manifest. 
> ```

Hey there @caheredia, love the work and thank you firstly. Question for you on the `ip_range` on the `vlan_net` network. Why do you only set a single ip address here? Wouldn't you need to set a range inclusive for the DNS, VPN, and Portainer services?

**Comment by caheredia** [comment]
@jaredtconnor I no longer use this setup. I'm running wireguard on pfsense. However, if I remember correctly I limited the dhcp `ip_range` to one IP address because I wanted to manually assign static addresses. And if I assigned a static address to a container that was within the `ip_range` addresses docker would complain. So this was kind of a workaround.

**Comment by tutods** [comment]
Hi everyone.
I have a macvlan for my **AdGuard Home**, and I can't access **AdGuard Home** using my **Wireguard** VPN.

Anyone can help me?
Thanks

**Comment by j0nnymoe** [comment]
@tutods this ticket has been closed for 2 years. What you're experiencing is by design of docker itself, I suggest reading up on what macvlan does and see how to work around it.
